File: ros2/camera_control/camera_control/camera/calibration.py
import argparse
import os
from glob import glob
from typing import List, Optional, Tuple

import cv2
import numpy as np


def calibrate_camera(
    mono_images: List[np.ndarray],
    grid_size: Tuple[int, int],
    grid_type: int = cv2.CALIB_CB_SYMMETRIC_GRID,
    blob_detector=None,
) -> Tuple[Optional[np.ndarray], Optional[np.ndarray]]:
    """
    Finds the camera intrinsic parameters and distortion coefficients from several views of a
    calibration pattern.

    Args:
        mono_images (List[np.ndarray]): Grayscale images each containing the calibration pattern.
        grid_size (Tuple[int, int]): (# cols, # rows) of the calibration pattern.
        grid_type (int): One of the following:
            cv2.CALIB_CB_SYMMETRIC_GRID uses symmetric pattern of circles.
            cv2.CALIB_CB_ASYMMETRIC_GRID uses asymmetric pattern of circles.
            cv2.CALIB_CB_CLUSTERING uses a special algorithm for grid detection. It is more robust to perspective distortions but much more sensitive to background clutter.
        blobDetector: Feature detector that finds blobs, like dark circles on light background. If None then a default implementation is used.

    Returns:
        Tuple[Optional[np.ndarray], Optional[np.ndarray]]: A tuple of (camera intrisic matrix, distortion coefficients), or (None, None) if calibration was unsuccessful.
    """
    # Prepare calibration pattern points,
    # These points are in the calibration pattern coordinate space. Since the calibration grid
    # is on a flat plane, we can set the Z coordinates as 0.
    calibration_points = np.zeros((grid_size[0] * grid_size[1], 3), np.float32)

    if grid_type == cv2.CALIB_CB_SYMMETRIC_GRID:
        calibration_points[:, :2] = np.mgrid[
            0 : grid_size[0], 0 : grid_size[1]
        ].T.reshape(-1, 2)
    elif grid_type == cv2.CALIB_CB_ASYMMETRIC_GRID:
        for i in range(0, grid_size[0] * grid_size[1]):
            row = i // grid_size[0]
            col = i % grid_size[0]
            row_offset = (row + 1) % 2
            x = col * 2 + row_offset
            y = row
            calibration_points[i] = (x, y, 0)
    else:
        raise Exception("Unsupported grid_type")

    obj_points = []
    img_points = []
    for image in mono_images:
        retval, centers = cv2.findCirclesGrid(
            image, grid_size, flags=grid_type, blobDetector=blob_detector
        )
        if retval:
            obj_points.append(calibration_points)
            img_points.append(centers)
        else:
            print(f"Could not get circle centers. Ignoring image.")

    try:
        retval, camera_matrix, dist_coeffs, rvecs, tvecs = cv2.calibrateCamera(
            obj_points, img_points, mono_images[0].shape[::-1], None, None
        )
        if retval:
            mean_error, errors = _calc_reprojection_error(
                obj_points, img_points, rvecs, tvecs, camera_matrix, dist_coeffs
            )
            print(
                f"Calibration successful. Used {len(obj_points)} images. Mean reprojection error: {mean_error}"
            )
            return camera_matrix, dist_coeffs
    except Exception as e:
        print(f"Error during calibration: {e}")

    print("Calibration unsuccessful.")
    return None, None


def _calc_reprojection_error(
    object_points: List[np.ndarray],
    image_points: List[np.ndarray],
    rvecs: Tuple[np.ndarray, ...],
    tvecs: Tuple[np.ndarray, ...],
    camera_matrix: np.ndarray,
    dist_coeffs: np.ndarray,
) -> Tuple[float, List[float]]:
    """
    Compute the reprojection error.

    Args:
        object_points (List[np.ndarray]): List of object points in real-world space
        image_points (List[np.ndarray]): List of corresponding image points detected in images
        rvecs (np.ndarray): List of rotation vectors returned by cv2.calibrateCamera
        tvecs (np.ndarray): List of translation vectors returned by cv2.calibrateCamera
        camera_matrix (np.ndarray): Camera matrix
        dist_coeffs (np.ndarray): Distortion coefficients

    Returns:
        float: Tuple of (mean reprojection error, list of per-image errors)
    """
    mean_error = 0
    total_points = 0
    errors = []
    for i in range(len(object_points)):
        projected_points, _ = cv2.projectPoints(
            object_points[i], rvecs[i], tvecs[i], camera_matrix, dist_coeffs
        )
        error = cv2.norm(image_points[i], projected_points, cv2.NORM_L2) / len(
            projected_points
        )
        errors.append(error)
        total_points += len(projected_points)
        mean_error += error

    mean_error /= len(object_points)
    return mean_error, errors


def create_blob_detector():
    """
    Blob detector for white circles on black background
    """
    params = cv2.SimpleBlobDetector_Params()

    # Filter by color
    params.filterByColor = True
    params.blobColor = 255

    # Filter by area
    params.filterByArea = True
    params.minArea = 10.0
    params.maxArea = 10000.0

    return cv2.SimpleBlobDetector_create(params)


def construct_extrinsic_matrix(rotation_vector, translation_vector):
    # Convert rotation vector to rotation matrix using Rodrigues' formula
    R, _ = cv2.Rodrigues(rotation_vector)

    # Create the extrinsic matrix
    extrinsic_matrix = np.eye(4)
    extrinsic_matrix[:3, :3] = R
    extrinsic_matrix[:3, 3] = translation_vector

    return extrinsic_matrix


def invert_extrinsic_matrix(extrinsic_matrix: np.ndarray):
    # Extract the rotation matrix and translation vector
    R = extrinsic_matrix[:3, :3]
    t = extrinsic_matrix[:3, 3]

    # Compute the inverse rotation matrix
    R_inv = R.T

    # Compute the new translation vector
    t_inv = -R_inv @ t

    # Construct the new extrinsic matrix
    extrinsic_matrix_inv = np.eye(4)
    extrinsic_matrix_inv[:3, :3] = R_inv
    extrinsic_matrix_inv[:3, 3] = t_inv

    return extrinsic_matrix_inv


def distort_pixel_coords(undistorted_pixel_coords, intrinsic_matrix, distortion_coeffs):
    # Convert the undistorted pixel coordinates to normalized camera coordinates
    fx, fy = intrinsic_matrix[0, 0], intrinsic_matrix[1, 1]
    cx, cy = intrinsic_matrix[0, 2], intrinsic_matrix[1, 2]

    normalized_coords = (undistorted_pixel_coords - [cx, cy]) / [fx, fy]
    normalized_coords = np.append(normalized_coords, [1.0])

    # Reshape to (1, 1, 3) for projectPoints input
    normalized_coords = normalized_coords.reshape(1, 1, 3)

    # Project the normalized coordinates to distorted pixel coordinates using projectPoints
    distorted_pixel_coords, _ = cv2.projectPoints(
        normalized_coords,
        np.zeros((3, 1)),
        np.zeros((3, 1)),
        intrinsic_matrix,
        distortion_coeffs,
    )

    # Extract the distorted pixel coordinates
    distorted_pixel_coords = distorted_pixel_coords[0, 0, :2]

    return distorted_pixel_coords


if __name__ == "__main__":
    parser = argparse.ArgumentParser(
        description="Calculate camera intrinsics and distortion coefficients from images of a calibration pattern"
    )
    parser.add_argument(
        "-i",
        "--input_dir",
        required=True,
        help="Path to the directory containing images of a calibration pattern",
    )

    args = parser.parse_args()
    image_paths = sorted(
        glob(os.path.join(args.input_dir, "*.jpg"))
        + glob(os.path.join(args.input_dir, "*.png"))
    )
    images = [
        cv2.cvtColor(cv2.imread(image_path), cv2.COLOR_RGB2GRAY)
        for image_path in image_paths
    ]
    camera_matrix, dist_coeffs = calibrate_camera(
        images,
        (5, 4),
        grid_type=cv2.CALIB_CB_SYMMETRIC_GRID,
        blob_detector=create_blob_detector(),
    )
    print(f"Calibrated intrins: {camera_matrix}")
    print(f"Distortion coeffs: {dist_coeffs}")

    # Test calibration
    img = images[10]
    h, w = img.shape[:2]
    newcameramtx, roi = cv2.getOptimalNewCameraMatrix(
        camera_matrix, dist_coeffs, (w, h), 1, (w, h)
    )
    undistorted_img = cv2.undistort(img, camera_matrix, dist_coeffs, None, newcameramtx)
    # crop the image
    x, y, w, h = roi
    undistorted_img = undistorted_img[y : y + h, x : x + w]
    cv2.imshow("undistorted", undistorted_img)
    cv2.waitKey(0)
    cv2.destroyAllWindows()

File: ros2/camera_control_cpp/src/camera/calibration.cpp
#include "camera_control_cpp/camera/calibration.hpp"

#include <filesystem>

#include "spdlog/spdlog.h"

std::optional<calibration::IntrinsicsResult> calibration::calculateIntrinsics(
    const std::vector<cv::Mat>& monoImages, const cv::Size& gridSize,
    const int gridType, const cv::Ptr<cv::FeatureDetector> blobDetector) {
  // Prepare calibration pattern points,
  // These points are in the calibration pattern coordinate space. Since the
  // calibration grid is on a flat plane, we can set the Z coordinates as 0.

  std::vector<cv::Point3f> calibrationPoints;
  if (gridType == cv::CALIB_CB_SYMMETRIC_GRID) {
    for (int i = 0; i < gridSize.height; i++) {
      for (int j = 0; j < gridSize.width; j++) {
        calibrationPoints.emplace_back(j, i, 0);
      }
    }
  } else if (gridType == cv::CALIB_CB_ASYMMETRIC_GRID) {
    for (int i = 0; i < gridSize.height; i++) {
      for (int j = 0; j < gridSize.width; j++) {
        calibrationPoints.emplace_back((2 * j + i % 2), i, 0);
      }
    }
  } else {
    spdlog::error("Unsupported grid type.");
    return std::nullopt;
  }

  if (monoImages.size() < 3) {
    spdlog::error(
        "At least 3 images are required to calculate the camera intrinsics.");
    return std::nullopt;
  }

  std::vector<std::vector<cv::Point3f>> objPoints;
  std::vector<std::vector<cv::Point2f>> imgPoints;
  bool found;
  for (const cv::Mat& image : monoImages) {
    std::vector<cv::Point2f> centers;
    found = findCirclesGrid(image, gridSize, centers, gridType, blobDetector);
    if (found) {
      objPoints.push_back(calibrationPoints);
      imgPoints.push_back(centers);
    } else {
      spdlog::warn("Could not get circle centers. Ignoring Image.");
    }
  }

  try {
    std::vector<cv::Mat> rvecs, tvecs;
    calibration::IntrinsicsResult result;
    result.intrinsicMatrix = cv::Mat::eye(3, 3, CV_64F);
    result.distCoeffs = cv::Mat::zeros(8, 1, CV_64F);
    float retval = cv::calibrateCamera(
        objPoints, imgPoints, monoImages[0].size(), result.intrinsicMatrix,
        result.distCoeffs, rvecs, tvecs);
    if (retval) {
      calibration::ReprojectErrors projErrors{
          calibration::_calcReprojectionError(objPoints, imgPoints, rvecs,
                                              tvecs, result.intrinsicMatrix,
                                              result.distCoeffs)};
      spdlog::info(
          "Calibration successful. Used {} images. Mean reprojection error: {}",
          objPoints.size(), projErrors.meanError);
      return result;
    }
  } catch (const std::exception& e) {
    spdlog::warn("Exception during calibration: \n{}", e.what());
  }

  spdlog::warn("Calibration unsuccessful");
  return std::nullopt;
}

cv::Mat calibration::constructExtrinsicMatrix(const cv::Mat& rvec,
                                              const cv::Mat& tvec) {
  cv::Mat R;

  // Convert rotation vector to rotation matric using Rodrigues' formula
  cv::Rodrigues(rvec, R);

  // Create extrinsic matrix
  cv::Mat extrinsic{cv::Mat::eye(4, 4, R.type())};
  R.copyTo(extrinsic(cv::Range(0, 3), cv::Range(0, 3)));
  tvec.copyTo(extrinsic(cv::Range(0, 3), cv::Range(3, 4)));

  return extrinsic;
}

std::pair<cv::Mat, cv::Mat> calibration::extractPoseFromExtrinsic(
    const cv::Mat& extrinsicMatrix) {
  // Extract rotation (3x3) and translation (3x1) from the 4x4 extrinsic matrix
  cv::Mat R{extrinsicMatrix(cv::Rect(0, 0, 3, 3)).clone()};     // top-left 3x3
  cv::Mat tvec{extrinsicMatrix(cv::Rect(3, 0, 1, 3)).clone()};  // 3x1 column

  // Convert rotation matrix to rotation vector (Rodrigues)
  cv::Mat rvec;
  cv::Rodrigues(R, rvec);

  return {rvec, tvec};
}

cv::Mat calibration::invertExtrinsicMatrix(const cv::Mat& extrinsic) {
  // Extract the rotation matrix and the translation vetor
  cv::Mat R{extrinsic(cv::Range(0, 3), cv::Range(0, 3))};
  cv::Mat t{extrinsic(cv::Range(0, 3), cv::Range(3, 4))};

  // Compute the inverse rotation matrix
  cv::Mat R_inv{R.t()};

  // Compute the new translation vector
  cv::Mat t_inv{-R_inv * t};

  // Construct the new extrinsic matrix
  cv::Mat extrinsic_inv{cv::Mat::eye(4, 4, R.type())};
  R_inv.copyTo(extrinsic_inv(cv::Range(0, 3), cv::Range(0, 3)));
  t_inv.copyTo(extrinsic_inv(cv::Range(0, 3), cv::Range(3, 4)));

  return extrinsic_inv;
}

std::optional<cv::Point2f> calibration::distortPixelCoords(
    const cv::Point2f& undistortedPixelCoords, const cv::Mat& intrinsicMatrix,
    const cv::Mat& distCoeffs) {
  if (intrinsicMatrix.type() != CV_64F) {
    spdlog::warn("Invalid type for Intrinsic Camera Matrix: {}",
                 intrinsicMatrix.type());
    return std::nullopt;
  }
  if (distCoeffs.type() != CV_64F) {
    spdlog::warn("Invalid type for Distortion Coefficients Matrix: {}",
                 distCoeffs.type());
    return std::nullopt;
  }

  // Extract focal length, principal point, etc.
  double fx{intrinsicMatrix.at<double>(0, 0)};
  double fy{intrinsicMatrix.at<double>(1, 1)};
  double cx{intrinsicMatrix.at<double>(0, 2)};
  double cy{intrinsicMatrix.at<double>(1, 2)};

  // Normalize Points & create 3d
  std::vector<cv::Point3f> normalizedPoints{
      {cv::Point3f((undistortedPixelCoords.x - cx) / fx,
                   (undistortedPixelCoords.y - cy) / fy, 1.0f)}};

  // Designate no rotation or translation
  cv::Mat rvec{cv::Mat::zeros(3, 1, CV_64F)};
  cv::Mat tvec{cv::Mat::zeros(3, 1, CV_64F)};

  // Project using distortion
  std::vector<cv::Point2f> distortedPoints;
  cv::projectPoints(normalizedPoints, rvec, tvec, intrinsicMatrix, distCoeffs,
                    distortedPoints);

  return distortedPoints[0];
}

cv::Mat calibration::scaleGrayscaleImage(const cv::Mat& monoImage) {
  // Ensure grayscale
  CV_Assert(monoImage.channels() == 1);

  // Convert to float to avoid overflow or underflow issues
  cv::Mat floatImg;
  monoImage.convertTo(floatImg, CV_32F);

  // Find the minimum and maximum values in the image
  double minVal, maxVal;
  cv::minMaxLoc(floatImg, &minVal, &maxVal);

  // Normalize the image to the range 0 to 1
  cv::Mat normalized;
  if (maxVal > minVal) {
    // Scale to [0,1]
    normalized = (floatImg - minVal) / (maxVal - minVal);
  } else {
    // Avoid division by zero, just shift
    normalized = floatImg - minVal;
  }

  // Scale to [0,255] as uint8
  cv::Mat scaled;
  normalized.convertTo(scaled, CV_8U, 255.0);

  return scaled;
}

cv::Ptr<cv::Feature2D> calibration::createBlobDetector() {
  cv::SimpleBlobDetector::Params params;

  // Filter By Color
  params.filterByColor = true;
  params.blobColor = 255;

  // Filter By Area
  params.filterByArea = true;
  params.minArea = 10.0;
  params.maxArea = 10000.0;

  return cv::SimpleBlobDetector::create(params);
}

calibration::ReprojectErrors calibration::_calcReprojectionError(
    const std::vector<std::vector<cv::Point3f>>& objectPoints,
    const std::vector<std::vector<cv::Point2f>>& imagePoints,
    const std::vector<cv::Mat>& rvecs, const std::vector<cv::Mat>& tvecs,
    const cv::Mat& cameraMatrix, const cv::Mat& distCoeffs) {
  calibration::ReprojectErrors retVals;
  float totalError{0};
  float err;
  for (size_t i = 0; i < objectPoints.size(); i++) {
    std::vector<cv::Point2f> projected;
    cv::projectPoints(objectPoints[i], rvecs[i], tvecs[i], cameraMatrix,
                      distCoeffs, projected);
    err = norm(imagePoints[i], projected, cv::NORM_L2) / projected.size();
    retVals.perImageErrors.push_back(err);
    totalError += err;
  }
  retVals.meanError = totalError / objectPoints.size();
  return retVals;
}

std::string calibration::expandUser(const std::string& path) {
  if (path.empty() || path[0] != '~') {
    return path;
  }

  const char* home{std::getenv("HOME")};
  if (!home) {
    throw std::runtime_error("HOME environment variable not set");
  }

  return std::string(home) + path.substr(1);
}

std::optional<cv::Mat> calibration::readXyzFile(const std::string& xyzFile) {
  std::filesystem::path filePath{calibration::expandUser(xyzFile)};
  if (!std::filesystem::exists(filePath)) {
    spdlog::error("File does not exist: {}", filePath.string());
    return std::nullopt;
  }
  cv::FileStorage fs{filePath, cv::FileStorage::READ};
  if (!fs.isOpened() || fs["xyz"].isNone()) {
    spdlog::error("Could not read XYZ data file: {}", filePath.string());
    return std::nullopt;
  }

  cv::Mat xyz;
  fs["xyz"] >> xyz;
  fs.release();

  return xyz;
}

std::optional<calibration::IntrinsicsResult> calibration::readIntrinsicsFile(
    const std::string& intrinsicsFile) {
  std::filesystem::path filePath{calibration::expandUser(intrinsicsFile)};
  if (!std::filesystem::exists(filePath)) {
    spdlog::error("Intrinsics file does not exist: {}", filePath.string());
    return std::nullopt;
  }
  cv::FileStorage fs{filePath, cv::FileStorage::READ};
  if (!fs.isOpened() || fs["intrinsicMatrix"].isNone() ||
      fs["distCoeffs"].isNone()) {
    spdlog::error("Invalid intrinsics file: {}", filePath.string());
    return std::nullopt;
  }

  cv::Mat intrinsicMatrix, distCoeffs;
  fs["intrinsicMatrix"] >> intrinsicMatrix;
  fs["distCoeffs"] >> distCoeffs;
  fs.release();

  calibration::IntrinsicsResult result;
  result.intrinsicMatrix = intrinsicMatrix;
  result.distCoeffs = distCoeffs;

  return result;
}

std::optional<cv::Mat> calibration::readExtrinsicsFile(
    const std::string& extrinsicsFile) {
  std::filesystem::path filePath{calibration::expandUser(extrinsicsFile)};
  if (!std::filesystem::exists(filePath)) {
    spdlog::error("Extrinsics file does not exist: {}", filePath.string());
    return std::nullopt;
  }
  cv::FileStorage fs{filePath, cv::FileStorage::READ};
  if (!fs.isOpened() || fs["extrinsicMatrix"].isNone()) {
    spdlog::error("Invalid extrinsics file: {}", filePath.string());
    return std::nullopt;
  }

  cv::Mat extrinsicMatrix;
  fs["extrinsicMatrix"] >> extrinsicMatrix;
  fs.release();

  return extrinsicMatrix;
}